### PR TITLE
add a class to create custom icons

### PR DESCRIPTION
**Pull request by devoncarew**
- add a class to create custom icons

This class creates small circular icons with a letter in the center. It uses the same style as the IntelliJ Structure view icons (also used a few other places in the UI). The indent here is to use this for something like the flutter inspector's widget tree, where you have a large number of elements, want the UI to fit in with IntelliJ, want some variety in the icons but don't want the user to remember hundreds of specific icons.

@jacob314 @pq @stevemessick

### CODE REVIEW

File: src/io/flutter/utils/CustomIconMaker.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import java.util.Map;
+
+public class CustomIconMaker {
+  private static final Color normalColor = ColorUtil.fromHex("231F20");
+
+  private final Map<String, Icon> iconCache = new HashMap<>();
+
+  public CustomIconMaker() {
+
+  }
```
Comment (pq) [inline src/io/flutter/utils/CustomIconMaker.java]: Nit: why the explicit default constructor?

File: src/io/flutter/utils/CustomIconMaker.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    return getCustomIcon(fromText, IconKind.kClass, false);
+  }
+
+  public Icon getCustomIcon(String fromText, IconKind kind) {
+    return getCustomIcon(fromText, kind, false);
+  }
+
+  public Icon getCustomIcon(String fromText, IconKind kind, boolean isAbstract) {
+    if (fromText == null || fromText.length() == 0) {
+      return null;
```
Comment (pq) [inline src/io/flutter/utils/CustomIconMaker.java]: Maybe `StringUtils.isEmpty(fromText)`?  Or:

    if (fromText == null || fromText.isEmpty()) { ...

File: src/io/flutter/utils/CustomIconMaker.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            g2.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
+            g2.setColor(normalColor);
+
+            final Font font = UIUtil.getFont(UIUtil.FontSize.MINI, UIUtil.getTreeFont());
+            g2.setFont(font);
+
+            final Rectangle2D bounds = g2.getFontMetrics().getStringBounds(text, g2);
+            final float offsetX = (getIconWidth() - (float)bounds.getWidth()) / 2.0f;
+            // Some black magic here for vertical centering.
+            final float offsetY = getIconHeight() - ((getIconHeight() - (float)bounds.getHeight()) / 2.0f) - 2.0f;
```
Comment (pq) [inline src/io/flutter/utils/CustomIconMaker.java]: Curious if you've tried this at different display resolutions?

File: src/io/flutter/utils/CustomIconMaker.java (hunk 1 above)
Comment (devoncarew) [reply src/io/flutter/utils/CustomIconMaker.java]: Hmm, habit? Should we avoid?

File: src/io/flutter/utils/CustomIconMaker.java (hunk 2 above)
Comment (devoncarew) [reply src/io/flutter/utils/CustomIconMaker.java]: done

File: src/io/flutter/utils/CustomIconMaker.java (hunk 3 above)
Comment (devoncarew) [reply src/io/flutter/utils/CustomIconMaker.java]: No (just HD), but this is similar to code that we do know works across resolutions.

File: src/io/flutter/utils/CustomIconMaker.java (hunk 1 above)
Comment (pq) [reply src/io/flutter/utils/CustomIconMaker.java]: Nope.  Just curious!

File: src/io/flutter/utils/CustomIconMaker.java (hunk 3 above)
Comment (pq) [reply src/io/flutter/utils/CustomIconMaker.java]: Cool.  Again, I was just curious!

**Pull request merged**